view failures from CI
```
=================================== FAILURES ===================================
_______________ TestRunCommand.test_failure_raises_runtime_error _______________

self = <tests.test_cli_provider.TestRunCommand object at 0x7f22e362da90>
mock_models = <MagicMock name='get_model_config' id='139787737267040'>
mock_run = <MagicMock name='run_cli' id='139787737267376'>

    @patch.dict("os.environ", {"KOAN_CLI_PROVIDER": "claude"})
    @patch("app.cli_exec.run_cli")
    @patch("app.config.get_model_config", return_value={"chat": "sonnet", "fallback": "haiku"})
    def test_failure_raises_runtime_error(self, mock_models, mock_run):
        """Non-zero exit raises RuntimeError with stderr snippet."""
        mock_run.return_value = MagicMock(returncode=1, stderr="some error message")
        with pytest.raises(RuntimeError, match="CLI invocation failed"):
>           run_command(
                prompt="analyze this",
                project_path="/fake/project",
                allowed_tools=["Read"],
            )

tests/test_cli_provider.py:1056: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
app/provider/__init__.py:232: in run_command
    result = run_cli_with_retry(
app/cli_exec.py:186: in run_cli_with_retry
    category = classify_cli_error(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

exit_code = 1, stdout = <MagicMock name='run_cli().stdout' id='139787736039504'>
stderr = 'some error message'

    def classify_cli_error(
        exit_code: int,
        stdout: str = "",
        stderr: str = "",
    ) -> ErrorCategory:
        """Classify a CLI error based on exit code and output text.
    
        Args:
            exit_code: Subprocess exit code (0 = success, not classified).
            stdout: Captured stdout from the CLI process.
            stderr: Captured stderr from the CLI process.
    
        Returns:
            ErrorCategory indicating how the caller should handle the error.
            Returns ``UNKNOWN`` for exit_code == 0 only if called explicitly
            (callers should not classify successful runs).
        """
        if exit_code == 0:
            return ErrorCategory.UNKNOWN
    
        combined = f"{stdout}\n{stderr}"
    
        # Check quota first — quota_handler is the authority for quota detection.
        # A 429 could be rate-limiting or quota exhaustion; defer to the
        # specialized detector which has provider-specific patterns.
        #
        # IMPORTANT: Use the same split-detection strategy as handle_quota_exhaustion
        # in quota_handler.py.  Loose patterns like "rate limit" and "too many
        # requests" can appear in Claude's stdout when it discusses API rate
        # limiting in its response text.  Only strict patterns are safe for stdout.
        from app.quota_handler import _STRICT_QUOTA_RE, _QUOTA_RE
    
>       if bool(_QUOTA_RE.search(stderr)) or bool(_STRICT_QUOTA_RE.search(stdout)):
                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       TypeError: expected string or bytes-like object, got 'MagicMock'

app/cli_errors.py:114: TypeError
=========================== short test summary info ============================
FAILED tests/test_cli_provider.py::TestRunCommand::test_failure_raises_runtime_error - TypeError: expected string or bytes-like object, got 'MagicMock'
=============== 1 failed, 9424 passed, 1607 deselected in 50.57s ===============
```

@sukria-koan0 rebase

--- a/koan/app/cli_errors.py
+++ b/koan/app/cli_errors.py
@@ -104,9 +104,14 @@ def classify_cli_error(
     # Check quota first — quota_handler is the authority for quota detection.
     # A 429 could be rate-limiting or quota exhaustion; defer to the
     # specialized detector which has provider-specific patterns.
-    from app.quota_handler import detect_quota_exhaustion
+    #
+    # IMPORTANT: Use the same split-detection strategy as handle_quota_exhaustion
+    # in quota_handler.py.  Loose patterns like "rate limit" and "too many
+    # requests" can appear in Claude's stdout when it discusses API rate
+    # limiting in its response text.  Only strict patterns are safe for stdout.
+    from app.quota_handler import _STRICT_QUOTA_RE, _QUOTA_RE
 
-    if detect_quota_exhaustion(combined):
+    if bool(_QUOTA_RE.search(stderr)) or bool(_STRICT_QUOTA_RE.search(stdout)):
         return ErrorCategory.QUOTA
 
     # Auth errors — Claude is logged out, needs human intervention.

--- a/koan/app/cli_errors.py
+++ b/koan/app/cli_errors.py
@@ -99,6 +99,10 @@ def classify_cli_error(
     if exit_code == 0:
         return ErrorCategory.UNKNOWN
 
+    # Coerce to strings — callers (and tests using MagicMock) may pass
+    # non-string values; regex search requires str input.
+    stdout = str(stdout) if stdout else ""
+    stderr = str(stderr) if stderr else ""
     combined = f"{stdout}\n{stderr}"
 
     # Check quota first — quota_handler is the authority for quota detection.

--- a/koan/tests/test_cli_errors.py
+++ b/koan/tests/test_cli_errors.py
@@ -208,3 +208,46 @@ class TestClassifyCliError:
         )
         result = classify_cli_error(1, stderr=stderr)
         assert result == ErrorCategory.AUTH
+
+    # -- False positive: loose quota patterns in stdout --------------------------
+
+    def test_no_false_positive_rate_limit_in_stdout(self):
+        """Loose patterns like 'rate limit' in stdout must NOT trigger QUOTA.
+
+        When Claude discusses API rate limiting in its response (stdout),
+        classify_cli_error should not confuse that with actual quota exhaustion.
+        Only strict patterns (e.g. 'out of extra usage') should match in stdout.
+        """
+        stdout = (
+            "Here's the plan for implementing rate limiting:\n"
+            "1. Add rate limit middleware to the API gateway\n"
+            "2. Configure per-endpoint rate limit thresholds\n"
+            "3. Return HTTP 429 with Retry-After header when limit exceeded"
+        )
+        result = classify_cli_error(1, stdout=stdout, stderr="Error: process crashed")
+        assert result != ErrorCategory.QUOTA, (
+            "Loose quota patterns in stdout caused false QUOTA classification"
+        )
+
+    def test_no_false_positive_usage_limit_in_stdout(self):
+        """'usage limit' in Claude's response should not trigger QUOTA."""
+        stdout = "You should set a usage limit on the API key to prevent abuse."
+        result = classify_cli_error(1, stdout=stdout, stderr="segfault")
+        assert result != ErrorCategory.QUOTA
+
+    def test_no_false_positive_too_many_requests_in_stdout(self):
+        """'too many requests' in Claude's code output should not trigger QUOTA."""
+        stdout = 'raise HTTPException(status_code=429, detail="too many requests")'
+        result = classify_cli_error(1, stdout=stdout, stderr="killed by signal")
+        assert result != ErrorCategory.QUOTA
+
+    def test_strict_patterns_still_match_in_stdout(self):
+        """Strict patterns like 'out of extra usage' should match even in stdout."""
+        stdout = "Error: out of extra usage quota for this billing period"
+        result = classify_cli_error(1, stdout=stdout)
+        assert result == ErrorCategory.QUOTA
+
+    def test_loose_patterns_match_in_stderr(self):
+        """Loose patterns should still match when they appear in stderr."""
+        result = classify_cli_error(1, stderr="rate limit exceeded")
+        assert result == ErrorCategory.QUOTA

